packages/flutter/lib/src/rendering/custom_paint.dart

```diff
@@ -0,0 +1,830 @@
+    Map<Key, SemanticsNode> oldKeyedChildren;
+    if (haveOldChildren) {
+      oldKeyedChildren = <Key, SemanticsNode>{};
+      while (oldChildrenTop <= oldChildrenBottom) {
+        final SemanticsNode oldChild = oldSemantics[oldChildrenTop];
+        if (oldChild.key != null)
+          oldKeyedChildren[oldChild.key] = oldChild;
+        oldChildrenTop += 1;
+      }
+    }
+
+    // Update the middle of the list.
+    while (newChildrenTop <= newChildrenBottom) {
+      SemanticsNode oldChild;
+      final CustomPainterSemantics newSemantics = newChildSemantics[newChildrenTop];
+      if (haveOldChildren) {
+        final Key key = newSemantics.key;
+        if (key != null) {
+          oldChild = oldKeyedChildren[key];
+          if (oldChild != null) {
+            if (_canUpdateSemanticsChild(oldChild, newSemantics)) {
+              // we found a match!
+              // remove it from oldKeyedChildren so we don't unsync it later
+              oldKeyedChildren.remove(key);
+            } else {
+              // Not a match, let's pretend we didn't see it for now.
+              oldChild = null;
+            }
+          }
+        }
+      }
+      assert(oldChild == null || _canUpdateSemanticsChild(oldChild, newSemantics));
+      final SemanticsNode newChild = _updateSemanticsChild(oldChild, newSemantics);
+      assert(oldChild == newChild || oldChild == null);
+      newChildren[newChildrenTop] = newChild;
+      newChildrenTop += 1;
+    }
+
+    // We've scanned the whole list.
+    assert(oldChildrenTop == oldChildrenBottom + 1);
+    assert(newChildrenTop == newChildrenBottom + 1);
+    assert(newChildSemantics.length - newChildrenTop == oldSemantics.length - oldChildrenTop);
+    newChildrenBottom = newChildSemantics.length - 1;
+    oldChildrenBottom = oldSemantics.length - 1;
+
+    // Update the bottom of the list.
+    while ((oldChildrenTop <= oldChildrenBottom) && (newChildrenTop <= newChildrenBottom)) {
+      final SemanticsNode oldChild = oldSemantics[oldChildrenTop];
+      final CustomPainterSemantics newSemantics = newChildSemantics[newChildrenTop];
+      assert(_canUpdateSemanticsChild(oldChild, newSemantics));
+      final SemanticsNode newChild = _updateSemanticsChild(oldChild, newSemantics);
+      assert(oldChild == newChild);
+      newChildren[newChildrenTop] = newChild;
+      newChildrenTop += 1;
+      oldChildrenTop += 1;
+    }
+
+    assert(() {
+      for (SemanticsNode node in newChildren) {
+        assert(node != null);
+      }
+      return true;
+    }());
+
+    return newChildren;
+  }
+
+  /// Whether `oldChild` can be updated with properties from `newSemantics`.
+  ///
+  /// If `oldChild` can be updated, it is updated using [_updateSemanticsChild].
+  /// Otherwise, the node is replaced by a new instance of [SemanticsNode].
+  static bool _canUpdateSemanticsChild(SemanticsNode oldChild, CustomPainterSemantics newSemantics) {
+    return oldChild.key == newSemantics.key;
```

Comment:
might be best to just inline this

Reply:
I'd like to keep this because it's used in 7 places and the name `_canUpdateSemanticsChild` makes the purpose of key comparison clearer than `if (oldChild.key == newChild.key)`.